qiskit/providers/experiment/analysis_result.py

```diff
@@ -0,0 +1,304 @@
+        """
+        self._source = result_data.pop(
+            "_source",
+            {"class": f"{self.__class__.__module__}.{self.__class__.__name__}",
+             "data_version": self._data_version,
+             "_qiskit_version": __qiskit_version__})
+
+        # Data to be stored in DB.
+        self._experiment_id = experiment_id
+        self._id = result_id or str(uuid.uuid4())
+        self._type = analysis_type
+        self._device_components = []
+        for comp in device_components:
+            if isinstance(comp, str):
+                comp = to_component(comp)
+            self._device_components.append(comp)
+        self._result_data = result_data
+        if isinstance(quality, str):
+            quality = ResultQuality(quality.upper())
+        self._quality = quality
+        self._quality_verified = verified
+        self._tags = tags or []
+        self._creation_date = datetime.now()
+
+        # Other attributes.
+        self._service = service
+        self._created_in_db = False
+        self.auto_save = False
+        if self._service:
+            try:
+                self.auto_save = self._service.option('auto_save')
+            except AttributeError:
+                pass
+        self._extra_data.update(kwargs)
+
+    def _serialize_data(self) -> str:
+        """Serialize result data into JSON string.
+
+        Returns:
+            Serialized JSON string.
+        """
+        return json.dumps(self._result_data, cls=self._json_encoder)
+
+    @classmethod
+    def deserialize_data(cls, data: str) -> Dict:
+        """Deserialize experiment from JSON string.
+
+        Args:
+            data: Data to be deserialized.
+
+        Returns:
+            Deserialized data.
+        """
+        return json.loads(data, cls=cls._json_decoder)
+
+    def save(
+            self,
+            service: Optional['ExperimentServiceV1'] = None
+    ) -> None:
+        """Save this analysis result in the database.
+
+        Args:
+            service: Experiment service to be used to save the data.
+                If ``None``, the default, if any, is used.
+
+        Raises:
+            ExperimentError: If the analysis result contains invalid data.
+        """
+        _result_data = json.loads(self._serialize_data())
+        _result_data['_source'] = self._source
+
+        new_data = {'experiment_id': self._experiment_id, 'result_type': self._type}
+        update_data = {'result_id': self.id, 'data': _result_data,
+                       'tags': self.tags, 'quality': self.quality}
+
+        service = service or self._service
+        if service:
+            self._created_in_db, _ = save_data(
+                is_new=self._created_in_db,
+                new_func=service.create_analysis_result,
+                update_func=service.update_analysis_result,
+                new_data=new_data, update_data=update_data)
+
+    def data(self) -> Dict:
```

Comment:
Maybe turn it into a property

Reply:
I kept `data()` and `tags()` as methods instead of properties to prevent people from do something like
```
result.data.update("new_key": "new_val")
```

^ this will not work with auto save. An alternative is to monitor updates to the `data` dictionary, as mentioned in the design doc. I eventually abandoned that idea because every update would generate an API call. If a user makes _n_ separate updates to `data` that's _n_ calls. This way if they do an `update_data()` in the end it's just 1 call.